fix: removes ability to have duplicate account names on front-end

Tigatok opened the pull request:
ref: https://github.com/actualbudget/actual/issues/316

<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes -->

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -49,7 +55,12 @@ export function CreateLocalAccountModal({
 
               const nameError = !name;
-              setNameError(nameError);
+              setNameError({ error: nameError, message: 'Name is required' });
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏we can put this in a conditional and return early if the error is encountered. Saves us a _tiny_ amount of compute.

```suggestion
              if (!name) {
                  setNameError({ error: true, message: 'Name is required' });
                  return;
              }
```

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -32,9 +33,14 @@ export function CreateLocalAccountModal({
-  const [nameError, setNameError] = useState(false);
+  const [nameError, setNameError] = useState({
+    error: false,
+    message: 'Name is required',
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏initially there should be no error
```suggestion
    message: '',
```

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -75,15 +86,17 @@ export function CreateLocalAccountModal({
                     const name = event.target.value.trim();
                     setName(name);
                     if (name && nameError) {
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏since we now have multiple types of errors: I think we can reset to the "no error" state if there previously was an error instead of using the `name` var
```suggestion
                    if (nameError.error) {
```

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -49,7 +55,12 @@ export function CreateLocalAccountModal({
-              setNameError(nameError);
+              setNameError({ error: nameError, message: 'Name is required' });
+
+              if (accounts.some(account => account.name === name)) {
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏I wonder if we should make this case insensitive. WDYT?

Tigatok replied:
I'd leave that up to the user. I'd say more control is better than less control, but that is just my opinion. LMK if we should change. Can always be done later.

MatissJanis reviewed (commented):
Thanks for working on this! Overall looks very good and solves 50% of the problem. The other 50% is: we should also catch this issue when an account is renamed.

Tigatok commented:
I changed a couple things, added the editing name error functionality as well to check for duplicated and add an error if found.
I also found/fixed a bug with the error state when using setState by using the local error value on the create account form. LMK if you want me to change anything else. 


https://github.com/actualbudget/actual/assets/3906461/896560af-3a71-4548-bdbe-7f99f1895964

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -48,13 +70,20 @@ export function CreateLocalAccountModal({
-              setNameError(nameError);
+              const nameError = validateName(name, accounts);
 
+              if (nameError.error) {
+                setNameError(nameError);
+              } else {
+                setNameError({ error: false, message: '' });
+              }
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏could simplify this further

```suggestion
              setNameError(nameError);
```

Tigatok replied:
I can take a look at that, I think when I tried that maybe it needed to get un-set, if not, I'm not sure why I didn't do that in the first place haha

MatissJanis commented on packages/desktop-client/src/components/modals/CreateLocalAccountModal.tsx:
```diff
@@ -23,6 +25,21 @@ type CreateLocalAccountProps = {
+    return { error: true, message: `Name: “${name}” already exists.` };
+  }
+
+  return { error: false, message: '' };
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏what if we always returned a `string` in this util? Then we could re-use it in both this modal and the account page. Slightly less duplication of logic.

psybers replied:
I agree.  If this just returns a string (empty string for no error) then you can reuse it above, and you can simplify the code below as you can just check `if (nameError)` instead of `if (nameError.error)`.  You can also apply the other suggestion Mattis has.

MatissJanis reviewed (commented):
Overall looks really good! Just two small comments. Let me know your thoughts.

joel-jeremy commented:
What do you guys think about putting the error message on the right of the input to avoid shifting the position of the transactions table?

Tigatok commented:
@joel-jeremy I did think of the position to the right, but I'm not too sure how that would work mobile-y. Maybe could look into a popup of some sort?

MatissJanis commented:
Good idea. A tooltip at the bottom with the error would fit nicely there IMO.

(if you use the tooltip component, then please use `common/Tooltip`)

Here's what we do for split transaction "errors". Not exactly a tooltip, but close-enough.
<img width="1214" alt="Screenshot 2024-05-22 at 17 37 36" src="https://github.com/actualbudget/actual/assets/886567/122fdcd2-149b-4107-b486-abe1702abfbe">

psybers commented:
Does this affect setting up bank sync?  Here is a scenario (based on real world):

I have 10 accounts that SimpleFIN lists with identical names (ugg...).  I want to import two of them from the dialog, so I select 'Create new account' next to both.

Currently, this would create two identically named accounts in Actual.  If I want to then disambiguate them, I could rename one/both.  Or, as in my case, maybe I don't really care and leave them the same because there will not be transfers to/from them.

psybers commented:
To answer my own question (without testing): I don't believe it does.  The bank sync code is a different path for adding accounts.
